### ISSUE TITLE
Fix View Switcher issue on Dataset Page

### ISSUE THREAD

**Issue by wilsaj** [open-austin/data-portal#31]
When editing a dataset, the form for editing is a bit off. Seems to be jumbled up with the Discus comment elements.

<img width="882" alt="screen shot 2016-07-28 at 8 19 18 pm" src="https://cloud.githubusercontent.com/assets/303866/17234717/013845ca-5501-11e6-982f-804edb695a1d.png">

**Comment by avickers** [comment]
The View Switcher is working as intended.  The issue is missing html attributes for some of the divs to indicate they shouldn't be shown when editing.

The spaghetti code that controls how this page works is pretty indecipherable to me.  You have to open nearly a dozen tabs to try to follow what's going on with this page.  But I think the solution can be pretty straightforward.

Add logic that only displays all of the errant divs when the `edit` query string isn't present.  I believe the way to accomplish that is by adding `data-hook="view" data-view="display"` to all of the outer divs of the components that need to be hidden when editing:
- `<div class="row" data-hook="view" data-view="display">
    <div class="col-xs-12 col-sm-6">
      <div class="exampletile" id = "datasetexampletile">`
- `<div class="col-xs-12" data-hook="view" data-view="display">
      <div class = "additional">
        <h2>Additional Info</h2>`
- `<div class = "disqus">
  <div id="disqus_thread"  data-hook="view" data-view="display"></div>`

Something like that.  I may also be the case that some of these require the `data-component="dataset-display"`.  It's very hard to follow Tim's flow.

**Comment by timwis** [comment]
Hey guys, have you read through the [components](https://github.com/timwis/jkan/wiki/Architecture#components) section of the wiki? The way it applies to the view switcher is as follows:
- The datasets collection is defined in [_config.yml](https://github.com/timwis/jkan/blob/gh-pages/_config.yml#L33-L41) and mapped to the `dataset` layout
- The [dataset layout](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html) contains a [view switcher component](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html#L14)
- The [first "view"](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html#L15) that can be switched (indicated by `data-hook="view"`) is called `display`
- The [second "view"](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html#L105) that can be switched is called `edit`
- `index.js` [instantiates](https://github.com/timwis/jkan/blob/gh-pages/scripts/src/index.js#L58) a `ViewSwitcher` class because layout has `data-component="view-switcher"`
- The `ViewSwitcher` class [shows](https://github.com/timwis/jkan/blob/gh-pages/scripts/src/components/view-switcher.js) the view specified in the querystring parameters. If there's no view specified, it just shows the first one.

So theoretically, `?view=display` would show the `display` view (displaying the dataset), and `?view=edit` would show the `edit` view. But since `display` is the first view, we don't ever need to actually have that in the URL.

So including disqus is just a matter of making sure the disqus code is inside the [`display` view](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html#L15) `<div>`. My hope was that the [comments at the top of `view-switcher.js`](https://github.com/timwis/jkan/blob/gh-pages/scripts/src/components/view-switcher.js#L7) would convey that.

I'd love your feedback on (a) how we could improve the wiki documentation to better convey this, and possibly (b) how the JavaScript architecture could be more simple and obvious. Ideally we'd have a separate page/layout for the `edit` view, but Jekyll only allows one page per collection item (until jekyll/jekyll#3041).

Hope this helps!

**Comment by avickers** [comment]
Hi Tim,

First, allow me to say that I think it's awesome that a City employee is creating projects like this and making them open source.  I'm not aware of anything like that going on here, so I give you mad props!  Please don't take anything that's about to follow personally.

I promise that I have read your documentation more than once.  The problem is that this is a project where I have a few hours to spend on it every few weeks.  My feedback would actually be to use a different approach:

This view switching / visibility / state issue is a very common problem, which means it's been addressed by just about every major framework and library.  I would strongly encourage you to adopt something like Angular, Knockout, React and use it instead of your own solution.

Your own solution is very professional and shows off your highly disciplined coding style, with a rigorous adherence to separation of concerns and DRY.  It is, however, singular and also quite possibly a million dollar solution to a ten dollar problem.  This is an application with 5 pages and two states.  Your solution using the three different data attributes would scale well to very complex applications, but this is a very simple application.

If you use a common framework, many developers will instantly understand your code upon seeing it the first time.  Moreover, they will still intuitively understand your code if they return to it months or years later.  Contrast that with my case, where I either have to reread your documentation or repuzzle my way through your code every few weeks, and the time I invest in doing this has no broader applicability.

These major frameworks are all well documented, they all have many tutorials, and they all have thousands upon thousands of Q&As on Stack Overflow to help the developer who gets stuck.  These are all tremendous assets.  Rolling your own solution creates a lot of needless technical debt for your project, even if that custom solution is well conceived and executed.  There's tremendous advantages and synergies to leveraging widely adopted Open Source libraries wherever possible.

**Comment by timwis** [comment]
hah, you make a good point, it may be a little overdone for the use case. I know the experience of having to relearn a codebase every once in a while. I would have used Backbone or a common framework, but had ruled those out because they're for single-page apps and this is multi-page: unless you write a separate JS file for each page, the JS file has no idea which page its on and what functionality it should enable. Previously, we had it checking the current route and enabling functionality based on the route. But the component approach [proposed](https://github.com/timwis/jkan/pull/87) a way to have components present on any page, letting users edit the Jekyll/Liquid markup to, say, add a search box, without having to necessarily touch the JS.

Having said that, I definitely understand your point and appreciate the constructive feedback. With the multi-page aspect in mind, do you still imagine there's a way to use a common framework?

**Comment by amaliebarras** [comment]
@timwis I've got basically this structure based on your recommendation above: 

```
<outer div w/ data-component="view-switcher">

<data-view="display">
dataset elements in here
disqus goes inside of here
</display view>

<data-view="edit">
</edit view>

</outer div>
```

Unfortunately it's still not working :( When I load the page, both the display AND edit views show up, and when I cancel out of the edit form, the view doesn't switch. 

I've also tried putting the edit view, inside of the display view, and that looks right on the page load, but then the edit form loads within the rest of the elements, below the fold, instead of them going away.  

<img width="1101" alt="screen shot 2016-08-09 at 8 56 17 pm" src="https://cloud.githubusercontent.com/assets/11826646/17539421/c2a2432a-5e73-11e6-998b-212a2d8a6a2f.png">

I'm trying to emulate the structure on [this](https://github.com/timwis/jkan/blob/gh-pages/_layouts/dataset.html#L14) page, but it's just not working the same for me!

Is @avickers right in that I need to put the <data-view="display"> on each element within the parent display div?

**Comment by timwis** [comment]
@amaliebarras is there a place I can see your code for `layouts/_dataset.html`? (Doesn't look like it's been pushed to this repo)

**Comment by timwis** [comment]
I've got it working in the upstream jkan `dataset.html` page [here](https://gist.github.com/timwis/d0cf9d015d9cb73ef7df5f43c22257ec#file-dataset-html-L103-L127) if that's helpful. We should totally make this part of jkan's default offering.

**Comment by amaliebarras** [comment]
@timwis My bad, just pushed it [here](https://github.com/open-austin/data-portal/blob/gh-pages/_layouts/dataset.html)

Agree on making disqus part of JKAN's default! I guess you'd have to put something in the config file so the user knows to set it up

**Comment by timwis** [comment]
@amaliebarras it looks like the disqus section is not quite contained inside the `display` `<div>`. There's a lot of HTML and nested logic on the page, so it can be hard to keep track of opening and closing tags. One thing that can help is to enforce indentation to make it a little more clear. Another thing we can do is take snippets out and put them into the `_includes` directory (perhaps even their own folder within it) and then insert them into the `dataset.html` page via `{% include snippet-name.html %}` similar to what's being done at the bottom with `dataset-form.html`. That would at least help readability.

I have your code running locally and I was going to try and submit a PR but there are a few other things going on here that I don't want to break -- it looks like there's a "ready to use" indicator and a "feedback" section, and I'm not sure whether you intend for the page to be a 2x2 grid or if that's accidental because of the markup. Is there a wireframe you're building off of? We can definitely get this going -- it's just an HTML markup issue.

**Comment by amaliebarras** [comment]
Indeed, there is definitely a lot going on here. I went over the opening and closing of tags so many times & I couldn't find it, so I'm glad you spotted it. I do appreciate the feedback on indentation / {% includes %} tags, should have done that to begin with, but I'll go back and fix it soon.

I actually do want that to be a 2x2. I don't have an up to date wireframe, unfortunately, I've changed my mind some since making mine. But let me take a look at the code again and try to figure out where that unclosed div is.
